File: ultimate/AbstractMessageFormatter.java
package ultimate;

public abstract class AbstractMessageFormatter implements MessageFormatter {

    public String formatMessage(String messageToFormat) {
        validateMessage();
        checkMessageSize();
        return format(messageToFormat);
    }

    private void checkMessageSize() {
        System.out.println("Checking message size...");
    }

    private void validateMessage() {
        System.out.println("validating the message....");
    }

    protected abstract String format(String messageToFormat);

    @Override
    public abstract String getType();
}

File: typescript/patterns/ultimate-design-pattern-challenge/abstract-message-formatter.ts
abstract class AbstractMessageFormatter implements MessageFormatter {
  formatMessage(messageToFormat: string): string {
    this.validateMessage(messageToFormat);
    this.checkMessageSize(messageToFormat);
    return this.format(messageToFormat);
  }

  private validateMessage(messageToValidate: string): void {
    console.log("Validating message...");
  }

  private checkMessageSize(messageToCheck: string): void {
    console.log("Checking message size...");
  }

  protected abstract format(messageToFormat: string): string;

  abstract getType(): string;
}
